hbase-server/src/main/java/org/apache/hadoop/hbase/security/provider/SaslServerAuthenticationProviders.java

```diff
@@ -0,0 +1,162 @@
+        providers = holder.get();
+        if (null != providers) {
+          return providers;
+        }
+
+        providers = createProviders(conf);
+        holder.set(providers);
+      }
+    }
+    return providers;
+  }
+
+  /**
+   * Removes the cached singleton instance of {@link SaslServerAuthenticationProviders}. 
+   */
+  public static void reset() {
+    synchronized (holder) {
+      holder.set(null);
+    }
+  }
+
+  /**
+   * Adds the given provider into the map of providers if a mapping for the auth code does not
+   * already exist in the map.
+   */
+  static void addProviderIfNotExists(SaslServerAuthenticationProvider provider,
+      HashMap<Byte,SaslServerAuthenticationProvider> providers) {
+    final byte newProviderAuthCode = provider.getSaslAuthMethod().getCode();
+    final SaslServerAuthenticationProvider alreadyRegisteredProvider = providers.get(
+        newProviderAuthCode);
+    if (alreadyRegisteredProvider != null) {
+      throw new RuntimeException("Trying to load SaslServerAuthenticationProvider "
+          + provider.getClass() + ", but "+ alreadyRegisteredProvider.getClass()
+          + " is already registered with the same auth code");
+    }
+    providers.put(newProviderAuthCode, provider);
+  }
+
+  /**
+   * Adds any providers defined in the configuration.
+   */
+  static void addExtraProviders(Configuration conf,
+      HashMap<Byte,SaslServerAuthenticationProvider> providers) {
+    for (String implName : conf.getStringCollection(EXTRA_PROVIDERS_KEY)) {
+      Class<?> clz;
+      try {
+        clz = Class.forName(implName);
+      } catch (ClassNotFoundException e) {
+        LOG.warn("Failed to find SaslServerAuthenticationProvider class {}", implName, e);
+        continue;
+      }
+
+      if (!SaslServerAuthenticationProvider.class.isAssignableFrom(clz)) {
+        LOG.warn("Server authentication class {} is not an instance of " 
+            + "SaslServerAuthenticationProvider", clz);
+        continue;
+      }
+
+      try {
+        SaslServerAuthenticationProvider provider =
+            (SaslServerAuthenticationProvider) clz.newInstance();
+        addProviderIfNotExists(provider, providers);
+      } catch (InstantiationException | IllegalAccessException e) {
+        LOG.warn("Failed to instantiate {}", clz, e);
+      }
+    }
+  }
+
+  /**
+   * Loads server authentication providers from the classpath and configuration, and then creates
+   * the SaslServerAuthenticationProviders instance.
+   */
+  static SaslServerAuthenticationProviders createProviders(Configuration conf) {
+    ServiceLoader<SaslServerAuthenticationProvider> loader =
+        ServiceLoader.load(SaslServerAuthenticationProvider.class);
+    HashMap<Byte,SaslServerAuthenticationProvider> providers = new HashMap<>();
+    for (SaslServerAuthenticationProvider provider : loader) {
+      addProviderIfNotExists(provider, providers);
+    }
+
+    addExtraProviders(conf, providers);
```

Comment:
The expected case is that including the provider on the classpath and creating the META-INF/services entry for SaslClientAuthenticationProvider (and SaslServerAuthenticationProvider) is sufficient.

However, to let unit tests work more easily (avoiding maven lifecycle issues and inherently insecure implementations from getting auto-loaded), you may also specify providers explicitly. The explicit configuration is more of a fail-safe switch, rather than the expected manner of us.